File: src/SlimShader/Chunks/Shex/SystemValueName.cs
using SlimShader.Util;

namespace SlimShader.Chunks.Shex
{
	public enum SystemValueName
	{
		[Description("NONE")]
		Undefined = 0,

		[Description("position")]
		Position = 1,

		[Description("clipdistance")]
		ClipDistance = 2,

		[Description("culldistance")]
		CullDistance = 3,

		[Description("rendertarget_array_index")]
		RenderTargetArrayIndex = 4,

		[Description("viewportarrayindex")]
		ViewportArrayIndex = 5,

		[Description("vertexid")]
		VertexID = 6,

		[Description("primitive_id")]
		PrimitiveID = 7,

		[Description("instanceid")]
		InstanceID = 8,

		[Description("isfrontface")]
		IsFrontFace = 9,

		[Description("sampleindex")]
		SampleIndex = 10,

		// The following are added for D3D11

		[Description("finalQuadUeq0EdgeTessFactor")]
		FinalQuadUEq0EdgeTessFactor = 11,

		[Description("finalQuadVeq0EdgeTessFactor")]
		FinalQuadVEq0EdgeTessFactor = 12,

		[Description("finalQuadUeq1EdgeTessFactor")]
		FinalQuadUEq1EdgeTessFactor = 13,

		[Description("finalQuadVeq1EdgeTessFactor")]
		FinalQuadVEq1EdgeTessFactor = 14,

		[Description("finalQuadUInsideTessFactor")]
		FinalQuadUInsideTessFactor = 15,

		[Description("finalQuadVInsideTessFactor")]
		FinalQuadVInsideTessFactor = 16,

		[Description("finalTriUeq0EdgeTessFactor")]
		FinalTriUEq0EdgeTessFactor = 17,

		[Description("finalTriVeq0EdgeTessFactor")]
		FinalTriVEq0EdgeTessFactor = 18,

		[Description("finalTriWeq0EdgeTessFactor")]
		FinalTriWEq0EdgeTessFactor = 19,

		[Description("finalTriInsideTessFactor")]
		FinalTriInsideTessFactor = 20,

		[Description("finalLineDetailTessFactor")]
		FinalLineDetailTessFactor = 21,

		[Description("finalLineDensityTessFactor")]
		FinalLineDensityTessFactor = 22
	}
}

File: src/cpp/Source/Chunks/Shex/SystemValueName.cpp
#include "PCH.h"
#include "SystemValueName.h"

using namespace std;
using namespace SlimShader;

string SlimShader::ToString(SystemValueName value)
{
	switch (value)
	{
	case SystemValueName::Undefined :
		return "NONE";
	case SystemValueName::Position :
		return "position";
	case SystemValueName::ClipDistance :
		return "clipdistance";
	case SystemValueName::CullDistance :
		return "culldistance";
	case SystemValueName::RenderTargetArrayIndex :
		return "rendertarget_array_index";
	case SystemValueName::ViewportArrayIndex :
		return "viewportarrayindex";
	case SystemValueName::VertexID :
		return "vertex_id";
	case SystemValueName::PrimitiveID :
		return "primitive_id";
	case SystemValueName::InstanceID :
		return "instanceid";
	case SystemValueName::IsFrontFace :
		return "isfrontface";
	case SystemValueName::SampleIndex :
		return "sampleindex";

	// The following are added for D3D11

	case SystemValueName::FinalQuadUEq0EdgeTessFactor :
		return "finalQuadUeq0EdgeTessFactor";
	case SystemValueName::FinalQuadVEq0EdgeTessFactor :
		return "finalQuadVeq0EdgeTessFactor";
	case SystemValueName::FinalQuadUEq1EdgeTessFactor :
		return "finalQuadUeq1EdgeTessFactor";
	case SystemValueName::FinalQuadVEq1EdgeTessFactor :
		return "finalQuadVeq1EdgeTessFactor";
	case SystemValueName::FinalQuadUInsideTessFactor :
		return "finalQuadUInsideTessFactor";
	case SystemValueName::FinalQuadVInsideTessFactor :
		return "finalQuadVInsideTessFactor";
	case SystemValueName::FinalTriUEq0EdgeTessFactor :
		return "finalTriUeq0EdgeTessFactor";
	case SystemValueName::FinalTriVEq0EdgeTessFactor :
		return "finalTriVeq0EdgeTessFactor";
	case SystemValueName::FinalTriWEq0EdgeTessFactor :
		return "finalTriWeq0EdgeTessFactor";
	case SystemValueName::FinalTriInsideTessFactor :
		return "finalTriInsideTessFactor";
	case SystemValueName::FinalLineDetailTessFactor :
		return "finalLineDetailTessFactor";
	case SystemValueName::FinalLineDensityTessFactor :
		return "finalLineDensityTessFactor";
	default :
		throw runtime_error("Unsupported value: " + to_string((int) value));
	}
}